lib/discourse_tagging.rb

```diff
@@ -794,4 +798,45 @@ def self.muted_tags(user)
     return [] unless user
     TagUser.lookup(user, :muted).joins(:tag).pluck("tags.name")
   end
+
+  def self.visible_tag_groups_for(guardian)
+    valid, uid, visible_tag_groups_names = Discourse.cache.read(VISIBLE_TAGGROUPS_CACHE_KEY)
+    if valid != true || uid != guardian.user&.id
+      # update cache
+      uid = guardian.user&.id
+      visible_tag_groups_names = TagGroup.visible(guardian).pluck(:name)
+      Discourse.cache.write(
+        VISIBLE_TAGGROUPS_CACHE_KEY,
+        [true, guardian.user&.id, visible_tag_groups_names],
+        expires_in: 1.minute,
+      )
+    end
+
+    visible_tag_groups_names
+  end
+
+  def self.tag_groups_for(tag)
+    # Record<String, Array<String> | [nil]>
+    tag_groups_map = Discourse.cache.read(TAGS_GROUPS_MAP_CACHE_KEY)
+    if !tag_groups_map
+      tag_groups_map =
+        Tag
+          .left_outer_joins(:tag_groups)
+          .select("tags.*, array_agg(tag_groups.name) as tag_group_names")
+          .group("tags.id")
+          .all
+          .each
+          .with_object({}) do |tag, hash|
+            hash[tag.name] = tag.tag_group_names
+            hash[tag.id] = tag.tag_group_names
+          end
```

Comment:
The number of tags on a site is unlimited. It wouldn't surprise me if we had sites with 100k+. 

So... I don't think we can afford to be holding them all in memory in a single object. It's likely to lead to performance issues on sites with lots of tags.

Reply:
Just curious... So what would /tags.json return for a forum with 100K+ tags?

Reply:
Yeah, we've had problems with that route on some sites, and will need to add pagination.

The good news is, in that case, it's a very rarely hit route. So even if it's slow, it doesn't affect the performance for most visitors.